src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.cs

```diff
@@ -0,0 +1,1244 @@
+        /// </summary>
+        /// <param name="left">The first ASCII text buffer to compare.</param>
+        /// <param name="right">The second ASCII text buffer to compare.</param>
+        /// <returns><see langword="true"/> if the buffers are equivalent; otherwise, <see langword="false"/>.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="right"/> is <see langword="null"/>.</exception>
+        public static bool EqualsIgnoreCase(ReadOnlySpan<byte> left, string right)
+        {
+            if (right is null)
+            {
+                throw new ArgumentNullException(nameof(right));
+            }
+
+            return EqualsIgnoreCase(left, right.AsSpan());
+        }
+
+        /*
+         * IndexOf routines
+         *
+         * Searches for the first (last) occurrence of the target substring within the search space,
+         * optionally treating [A-Z] and [a-z] as equal. All non-ASCII bytes are compared for pure
+         * binary equivalence. Returns the index of where the first (last) match is found, else returns -1.
+         */
+
+        /// <summary>
+        /// Given two ASCII text buffers, searches for the first occurrence of where
+        /// <paramref name="value"/> appears within <paramref name="text"/>.
+        /// </summary>
+        /// <param name="text">The ASCII text buffer which will be searched.</param>
+        /// <param name="value">The ASCII text which will be sought within <paramref name="text"/>.</param>
+        /// <returns>The zero-based index in <paramref name="text"/> where <paramref name="value"/>
+        /// first appears; else -1 if <paramref name="value"/> is not found within <paramref name="text"/>.</returns>
+        public static int IndexOf(ReadOnlySpan<byte> text, ReadOnlySpan<byte> value)
+            => text.IndexOf(value);
+
+        /// <summary>
+        /// Given two ASCII text buffers, searches for the first occurrence of where
+        /// <paramref name="value"/> appears within <paramref name="text"/> using a
+        /// case-insensitive comparison.
+        /// </summary>
+        /// <param name="text">The ASCII text buffer which will be searched.</param>
+        /// <param name="value">The ASCII text which will be sought within <paramref name="text"/>.</param>
+        /// <returns>The zero-based index in <paramref name="text"/> where <paramref name="value"/>
+        /// first appears; else -1 if <paramref name="value"/> is not found within <paramref name="text"/>.</returns>
+        public static int IndexOfIgnoreCase(ReadOnlySpan<byte> text, ReadOnlySpan<byte> value)
+        {
+            // If 'value' is longer than 'text', 'lastIdx' will go negative.
+            // That's acceptable since it'll early exit the loop below.
+
+            int lastIdx = text.Length - value.Length;
+            for (int i = 0; i <= lastIdx; i++)
+            {
+                if (EqualsIgnoreCase(text.Slice(i, value.Length), value))
```

Comment:
This seems inefficient.  In the majority case I'd expect the text not to match at any given location, and likely even for the first character not to match, but we're slicing before we're able to do any checks, and we're validating the length at every position even though we know it's going to match.  Does the JIT successfully eliminate a lot of that overhead?  I know when I was looking at regex optimizations, and I tried replacing certain aspects with a pattern that looked very much like this, perf dropped measurably.

Reply:
I see the ideal implementation of this method being: check if the right-hand side contains only a-zA-Z. If this is the case (which I suspect will be common), then clone the logic of MemoryExtensions.IndexOf but include an extra "OR all intermediate values by 0x20" step. Symbols will be corrupted in the intermediate registers by this, but that's ok since we weren't looking for them anyway.

Reply:
@stephentoub has a very good point here 👍